server/api/impl/admin/synchronize.go

```diff
@@ -180,7 +113,127 @@ func (adm *Admin) UninstallApp(appID apps.AppID) error {
+// appsForRegistration are apps from the apps.json file
+// registeredApps are apps in the config already registered.
+func mergeApps(appsForRegistration apps.AppVersionMap, registeredApps []*apps.App) (apps.AppVersionMap, map[apps.AppID]appOldVersion, map[apps.AppID]*apps.App) {
+	appsToRegister := apps.AppVersionMap{}
+	appsToUpgrade := map[apps.AppID]appOldVersion{}
+	appsToRemove := map[apps.AppID]*apps.App{}
+
+	allRegisteredAppsMap := map[apps.AppID]*apps.App{}
+	for _, registeredApp := range registeredApps {
+		allRegisteredAppsMap[registeredApp.AppID] = registeredApp
+		newVersion, ok := appsForRegistration[registeredApp.AppID]
+		if !ok {
+			// app was registered but it is not in the registration list anymore.
+			appsToRemove[registeredApp.AppID] = registeredApp
+			continue
+		}
+		if newVersion != registeredApp.Manifest.Version {
+			appsToUpgrade[registeredApp.AppID] = appOldVersion{
+				oldApp:     registeredApp,
+				newVersion: newVersion,
+			}
+		}
+	}
+	for id, appToRegister := range appsForRegistration {
+		if _, ok := allRegisteredAppsMap[id]; !ok {
+			// here is a new app
+			appsToRegister[id] = appToRegister
+		}
+	}
+	return appsToRegister, appsToUpgrade, appsToRemove
+}
+
+func (adm *Admin) removeApps(appsToRemove map[apps.AppID]*apps.App) {
+	for _, appToRemove := range appsToRemove {
+		switch appToRemove.Status {
+		case apps.AppStatusInstalled:
+			if err := adm.UninstallApp(appToRemove.AppID); err != nil {
+				adm.mm.Log.Error("can't uninstall app", "app_id", appToRemove.AppID, "err", err.Error())
+			}
+		case apps.AppStatusRegistered:
+			// delete app from proxy plugin, not removing the data
+			if err := adm.store.App().Delete(appToRemove); err != nil {
+				adm.mm.Log.Error("can't delete app from store", "app_id", appToRemove.AppID, "err", err.Error())
+			}
+		}
+	}
+}
+
+func (adm *Admin) upgradeApps(appsToUpgrade map[apps.AppID]appOldVersion) map[apps.AppID]appNewVersion {
+	upgradedApps := map[apps.AppID]appNewVersion{}
+	for _, appToUpgrade := range appsToUpgrade {
+		oldVersion := appToUpgrade.oldApp.Manifest.Version
+		newManifest, err := adm.store.Manifest().Get(appToUpgrade.oldApp.AppID)
+		if err != nil {
+			adm.mm.Log.Error("can't load manifest from store", "app_id", appToUpgrade.oldApp.AppID, "err", err.Error())
+			continue
+		}
+		if newManifest.Version != appToUpgrade.newVersion {
+			adm.mm.Log.Error("versions do not match this should not happen", "app_id", appToUpgrade.oldApp.AppID)
+			continue
+		}
+		upgradedApp := appToUpgrade.oldApp
+		upgradedApp.Manifest = newManifest
+		if err := adm.store.App().Save(upgradedApp); err != nil {
+			adm.mm.Log.Error("can't save an app", "app_id", upgradedApp.AppID, "err", err.Error())
+			continue
+		}
+		upgradedApps[upgradedApp.AppID] = appNewVersion{
+			newApp:     upgradedApp,
+			oldVersion: oldVersion,
+		}
+	}
+	return upgradedApps
+}
+
+func (adm *Admin) registerApps(appsToRegister apps.AppVersionMap) {
```

Comment:
Should we use the term `listed App` instead? Register indicates to me that the App did register itself, which is not the case. The term `listed` is also used in the Marketplace for the same kind of Apps as here.

Reply:
I'm 0/5 on naming. IT was called `listed` in my previous PR but we changed it to `registered`.

Reply:
Moving to https://github.com/mattermost/mattermost-plugin-apps/pull/85